packages/shorebird_cli/lib/src/executables/devicectl/devicectl.dart

```diff
@@ -0,0 +1,311 @@
+import 'package:shorebird_cli/src/third_party/flutter_tools/lib/flutter_tools.dart';
+import 'package:shorebird_code_push_client/shorebird_code_push_client.dart';
+
+typedef BundleId = String;
+
+/// {@template devicectl_exception}
+/// Thrown when a [Devicectl] command fails.
+/// {@endtemplate}
+class DevicectlException implements Exception {
+  /// {@macro devicectl_exception}
+  DevicectlException({
+    required this.message,
+    this.underlyingException,
+  });
+
+  /// A message describing this exception.
+  final String message;
+
+  /// The exception that caused this exception to be thrown, if any.
+  final Exception? underlyingException;
+
+  @override
+  String toString() => '''
+DevicectlException: $message
+Underlying exception: ${underlyingException ?? '(none)'}
+''';
+}
+
+/// A reference to a [Devicectl] instance.
+final devicectlRef = create(Devicectl.new);
+
+/// The [Devicectl] instance available in the current zone.
+Devicectl get devicectl => read(devicectlRef);
+
+/// A wrapper around the `devicectl` command.
+class Devicectl {
+  static const executableName = 'xcrun';
+  static const baseArgs = [
+    'devicectl',
+  ];
+
+  /// Whether the `devicectl` command is available.
+  Future<bool> _isAvailable() async {
+    try {
+      final result = await process.run(executableName, [
+        ...baseArgs,
+        '--version',
+      ]);
+      return result.exitCode == ExitCode.success.code;
+    } catch (_) {
+      return false;
+    }
+  }
+
+  /// Returns the first available iOS device, or the device with the given
+  /// [deviceId] if provided. Devices that are running iOS <17 are not
+  /// "CoreDevice"s and are not visible to devicectl.
+  Future<AppleDevice?> _deviceForLaunch({String? deviceId}) async {
+    final devices = await listAvailableIosDevices();
+
+    if (deviceId != null) {
+      return devices.firstWhereOrNull((d) => d.identifier == deviceId);
+    } else {
+      return devices.firstOrNull;
+    }
+  }
+
+  /// Whether we should use `devicectl` to install and launch the app on the
+  /// device with the given [deviceId], or the first available device we find if
+  /// [deviceId] is not provided.
+  Future<bool> shouldUseDevicectl({String? deviceId}) async {
+    if (!await _isAvailable()) {
+      return false;
+    }
+
+    return await _deviceForLaunch(deviceId: deviceId) != null;
+  }
+
+  /// Installs the given [runnerApp] on the device with the given [deviceId].
+  ///
+  /// Returns the bundle ID of the installed app.
+  Future<BundleId> installApp({
+    required Directory runnerApp,
+    required String deviceId,
+  }) async {
+    const failureErrorMessage = 'App install failed';
+
+    final args = [
+      ...baseArgs,
+      'device',
+      'install',
+      'app',
+      '--device',
+      deviceId,
+      runnerApp.path,
+    ];
+    final Json jsonResult;
+    try {
+      jsonResult = await _runJsonCommand(args: args);
+    } catch (error) {
```

Comment:
```suggestion
    } on Exception catch (error) {
```

so we don't need to cast below

Reply:
Done